File: app/src/main/java/com/example/wangbin/gymclub/RegisterActivity.java
package com.example.wangbin.gymclub;

import android.content.Intent;
import android.os.Bundle;
import android.os.Looper;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import com.example.wangbin.gymclub.net.HttpCallbackListener;
import com.example.wangbin.gymclub.net.HttpSettings;
import com.example.wangbin.gymclub.net.HttpUtil;

import org.json.JSONObject;

import java.util.HashMap;

public class RegisterActivity extends AppCompatActivity {
    private EditText mId;
    private EditText mPhone;
    private EditText mPassword,mPassword2;
    private EditText mName;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);
        ImageButton mEmailView = (ImageButton) findViewById(R.id.imageButton11);
        Button singupButton = (Button) findViewById(R.id.account_sign_up_button);
        mEmailView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //attemptLogin();
                finish();
            }
        });
        mId= (EditText) findViewById(R.id.userid);
        mPhone= (EditText) findViewById(R.id.phone);
        mName= (EditText) findViewById(R.id.name);
        mPassword= (EditText) findViewById(R.id.password1);
        mPassword2= (EditText) findViewById(R.id.password2);

        singupButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String id = mId.getText().toString();
                String pwd = mPassword.getText().toString();
                String phone = mPhone.getText().toString();
                String name = mName.getText().toString();
                String pwd2 = mPassword2.getText().toString();
                if (id.length() <= 4 || id.length() > 20) {
                    Toast toast = Toast.makeText(RegisterActivity.this, "账号输入应该在5-20位", Toast.LENGTH_SHORT);
                    mId.setText("");
                    toast.show();
                } else if (pwd.length() <= 5 || pwd.length() > 20) {
                    Toast toast = Toast.makeText(RegisterActivity.this, "密码输入应该在6-20位", Toast.LENGTH_SHORT);
                    mPassword.setText("");
                    toast.show();
                } else if (name.length()<=1) {
                    Toast toast = Toast.makeText(RegisterActivity.this, "姓名输入错误", Toast.LENGTH_SHORT);
                    mName.setText("");
                    toast.show();
                } else if (pwd2.length() <= 5 || pwd2.length() > 20) {
                    Toast toast = Toast.makeText(RegisterActivity.this, "确认密码输入应该在6-20位", Toast.LENGTH_SHORT);
                    mPassword2.setText("");
                    toast.show();
                } else if (phone.length() !=11) {
                    Toast toast = Toast.makeText(RegisterActivity.this, "手机输入应该为11位", Toast.LENGTH_SHORT);
                    mPhone.setText("");
                    toast.show();
                } else if (!pwd.equals(pwd2)) {
                    Toast toast = Toast.makeText(RegisterActivity.this, "确认密码与密码输入不一致", Toast.LENGTH_SHORT);
                    mPassword.setText("");
                    mPassword2.setText("");
                    toast.show();
                } else {
                    HashMap<String, String> map = new HashMap<String, String>();
                    map.put("user_id", id);
                    map.put("user_pwd", pwd);
                    map.put("user_phone", phone);
                    map.put("user_name", name);
                    HttpUtil.sendHttpPostRequest(HttpSettings.httpUrl + HttpSettings.registerUrl, map, new HttpCallbackListener() {
                        @Override
                        public void onFinish(String response) {
                            try {
                                JSONObject res = new JSONObject(response);
                                String result = res.getString("result");
                                if (result.equals("true")) {
                                    JSONObject data = res.getJSONObject("data");
                                    Intent i = new Intent(RegisterActivity.this, MainActivity.class);
                                    i.putExtra("name", data.getString("name"));
                                    i.putExtra("phone", data.getString("phone"));
                                    startActivity(i);
                                } else {
                                    Looper.prepare();
                                    Toast toast = Toast.makeText(RegisterActivity.this, "账号或手机已经被注册", Toast.LENGTH_SHORT);
                                    mId.setText("");
                                    mPhone.setText("");
                                    toast.show();
                                    Looper.loop();
                                }

                            } catch (Exception e) {
                                Looper.prepare();
                                Toast toast = Toast.makeText(RegisterActivity.this, "注册失败", Toast.LENGTH_SHORT);
                                mId.setText("");
                                mPhone.setText("");
                                mName.setText("");
                                mPassword.setText("");
                                mPassword2.setText("");
                                toast.show();
                                Looper.loop();
                            }

                        }
                        @Override
                        public void onError(Exception e) {
                            Looper.prepare();
                            Toast toast = Toast.makeText(RegisterActivity.this, "注册失败,请检查网络", Toast.LENGTH_SHORT);
                            toast.show();
                            Looper.loop();
                        }
                    });

                }
            }
        });
    }
}

File: app/src/main/java/com/example/wangbin/gymclub/RegisterActivity.kt
package com.example.wangbin.gymclub

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.Toast
import com.example.wangbin.gymclub.net.RetrofitUtils
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_register.*
import okhttp3.*
import java.io.IOException
import java.util.concurrent.TimeUnit

class RegisterActivity : Activity() {
    private var mPhoneNum: String? = null
    private var mobile_flag: Boolean = false
    private var password_flag: Boolean = false
    private var password:String? = null
    private var phoneNum:String? = null
    private var disposable:Disposable? = null

    inline fun <reified T : Activity> Context.startActivity() {
        startActivity(Intent(this, T::class.java))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
        close.setOnClickListener({finish()})
        et_mobile.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
                phoneNum = s.toString()
                s?.let {
                    mPhoneNum = s.toString()
                    if (it.length == 11) {
                        mobile_flag = true
                        update()
                    } else {
                        mobile_flag = false
                        update()
                    }
                }
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {

            }

        })
        et_password.addTextChangedListener(object : TextWatcher{
            override fun afterTextChanged(s: Editable?) {
                password = s.toString()
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                if (et_password.text != null && !et_password.text.toString().equals("") && et_password_again.text != null
                        && !et_password_again.text.toString().equals("")) {
                    if (et_password_again.text.toString().equals(et_password.text.toString())) {
                        password_flag = true
                        update()
                    } else {
                        password_flag = false
                        update()
                    }
                } else {
                    password_flag = false
                    update()
                }
            }

        })
        et_password_again.addTextChangedListener(object  : TextWatcher{
            override fun afterTextChanged(s: Editable?) {

            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                if (et_password.text != null && !et_password.text.toString().equals("") && et_password_again.text != null
                        && !et_password_again.text.toString().equals("")) {
                    if (et_password_again.text.toString().equals(et_password.text.toString())) {
                        password_flag = true
                        update()
                    } else {
                        password_flag = false
                        update()
                    }
                } else {
                    password_flag = false
                    update()
                }
            }

        })
        btn_register.setOnClickListener({
            var register_result: Boolean = false
//            var okHttpClient: OkHttpClient = OkHttpClient.Builder().connectTimeout(5000, TimeUnit.MILLISECONDS).readTimeout(5000, TimeUnit.MILLISECONDS).build();
//            var requestBody:RequestBody = FormBody.Builder().add("phonenum",phoneNum)
//                    .add("password",password)
//                    .build()
//            var request : Request = Request.Builder().url("http://172.27.141.31:8110/register").post(requestBody).build()
//            okHttpClient.newCall(request).enqueue(object : Callback{
//                override fun onFailure(call: Call?, e: IOException?) {
//
//                }
//
//                override fun onResponse(call: Call?, response: Response?) {
//                    Log.d("SSS",response?.body()?.string())
//                }
//
//            })
//        })
            if (et_password.text.toString().equals(et_password_again.text.toString())) {
                RetrofitUtils.register(phoneNum, password)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(object : Observer<ResponseBody> {
                            override fun onComplete() {
                                if (register_result == true) {
                                    startActivity<LoginActivity>()
                                    finish()
                                }else{
                                    Toast.makeText(this@RegisterActivity,R.string.error_register,Toast.LENGTH_SHORT).show()
                                }
                            }

                            override fun onSubscribe(d: Disposable) {
                                disposable = d
                            }

                            override fun onNext(t: ResponseBody) {
                                register_result = t.string().toBoolean()
                            }

                            override fun onError(e: Throwable) {
                                e.printStackTrace()
                            }

                        })
            }
        })

    }

    override fun onDestroy() {
        super.onDestroy()
        disposable?.dispose()
    }

    fun update() {
        if (!password_flag) {
            if (et_password.text != null && !et_password.text.toString().equals("") && et_password_again.text != null
                    && !et_password_again.text.toString().equals("")) {
                error_message.visibility = View.VISIBLE
            } else {
                error_message.visibility = View.GONE
            }
        } else {
            error_message.visibility = View.GONE
        }
        if (mobile_flag && et_password_again.text != null && !et_password_again.text.toString().equals("")
                && et_password.text != null && !et_password.text.toString().equals("")) {
            btn_register.isEnabled = true
        } else {
            btn_register.isEnabled = false
        }
    }
}